src/Razor/src/Microsoft.CodeAnalysis.Razor.Workspaces/AdhocWorkspaceProvider.cs

```diff
@@ -0,0 +1,35 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT license. See License.txt in the project root for license information.
+
+using System;
+
+namespace Microsoft.CodeAnalysis.Razor.Workspaces;
+
+internal sealed class AdhocWorkspaceProvider(IHostServicesProvider hostServicesProvider) : IDisposable
+{
+    private readonly Lazy<AdhocWorkspace> _lazyWorkspace = new(() => CreateWorkspace(hostServicesProvider));
+
+    private static AdhocWorkspace CreateWorkspace(IHostServicesProvider hostServicesProvider)
+    {
+        var fallbackServices = hostServicesProvider.GetServices();
+        var services = AdhocServices.Create(
+            workspaceServices: [],
+            languageServices: [],
+            fallbackServices);
+
+        return new AdhocWorkspace(services);
+    }
+
+    public AdhocWorkspace GetOrCreate()
+    {
+        return _lazyWorkspace.Value;
+    }
+
+    public void Dispose()
+    {
+        if (_lazyWorkspace.IsValueCreated)
+        {
+            _lazyWorkspace.Value.Dispose();
+        }
```

Comment:
The disposal contract of this class is a little squishy. Essentially, it can be disposed multiple times until the workspace is created? I'm noting this here because I found a couple of places in tests where a `FormattingContext` to be created with an `AdhocWorkspaceProvider` that is disposed before the context is used. There's no disposable after the `FormattingContext` is used, so the workspace in the `AdhocWorkspaceProvider` would never be disposed.